dask/array/routines.py

```diff
@@ -929,6 +932,304 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
+        # We have integer bins and tuple ranges
+        if all(isinstance(b, int) for b in bins) and all(len(r) == 2 for r in range):
+            edges = [np.linspace(r[0], r[1], b + 1) for b, r in zip(bins, range)]
+        # We have arrays of bin edges
+        else:
+            edges = [np.asarray(b) for b in bins]
+        # at this point we've converted bins + range combinations to
+        # edge arrays, so we can null the range argument later passed
+        # to the callable in the graph (_block_histogramdd).
+        range = (None,) * D
+    else:
+        raise NotImplementedError(
+            "Passing dask collections to bins=... or range=... is not supported."
+        )
+
+    # This call is for potential compatibility with delayed bin
+    # configurations. For non-delayed bin confurations this returns
+    # the concrete objects and deps is an empty tuple.
+    (bins_ref, ranges_ref), deps = unpack_collections([edges, range])
+
+    # With dsk below, we will construct a (D + 1) dimensional array
+    # stacked for each chunk. For example, if the histogram is going
+    # to be 3 dimensions, this creates a stack of cubes (1 cube for
+    # each sample chunk) that will be collapsed into a final cube (the
+    # result).
+
+    # This tuple of zeros represents the chunk index along the columns
+    # (we only allow chunking along the rows).
+    column_zeros = tuple(0 for _ in _range(D))
+
+    if weights is None:
+        dsk = {
+            (name, i, *column_zeros): (_block_histogramdd, k, bins_ref, ranges_ref)
+            for i, k in enumerate(flatten(sample.__dask_keys__()))
+        }
+        dtype = np.histogramdd([])[0].dtype
+    else:
+        a_keys = flatten(sample.__dask_keys__())
+        w_keys = flatten(weights.__dask_keys__())
+        dsk = {
+            (name, i, *column_zeros): (_block_histogramdd, k, bins_ref, ranges_ref, w)
+            for i, (k, w) in enumerate(zip(a_keys, w_keys))
+        }
+        dtype = weights.dtype
+
+    deps = (sample,) + deps
+    if weights is not None:
+        deps += (weights,)
+    graph = HighLevelGraph.from_collections(name, dsk, dependencies=deps)
+
+    nchunks = len(list(flatten(sample.__dask_keys__())))
+    all_nbins = tuple((b.size - 1,) for b in edges)
+    stacked_chunks = ((1,) * nchunks, *all_nbins)
+    mapped = Array(graph, name, stacked_chunks, dtype=dtype)
+
+    # Finally, sum over chunks providing to get the final D
+    # dimensional result array.
+    n = mapped.sum(axis=0)
+
+    if density:
+        # compute array of values to divide by the bin width along
+        # each dimension.
+        width_divider = np.ones(n.shape)
+        for i in _range(D):
+            shape = np.ones(D, int)
+            shape[i] = width_divider.shape[i]
+            width_divider *= np.diff(edges[i]).reshape(shape)
+        width_divider = asarray(width_divider, chunks=n.chunks)
+        return n / width_divider / n.sum(), edges
```

Comment:
Would it be possible to handle this within `_block_histogramdd`?

Reply:
I thought about this for a bit and I can't think of useful way. If we compute the density in each block we'd have to track the fraction of each block's contribution to the final aggregated histogram. I imagine it's possible but I think that would be a more complex implementation (and it gets more even more complicated when bits of the data in each chunk fall outside of the axis range(s)).

Reply:
Yeah was just thinking of how we can get these operations fused effectively. Though I think if complicates understanding and maintaining the code, it's probably not worth it. Dask already has some logic to fuse things on its own. We can always revisit if it's an issue. Thanks for looking into this 🙂